configure.ac

```diff
@@ -2178,7 +2178,6 @@ AC_MSG_NOTICE([X-Accelerator-Vary support enabled: $enable_x_accelerator_vary])
 
 SQUID_AUTO_LIB(cppunit,[cppunit test framework],[LIBCPPUNIT])
 SQUID_CHECK_LIB_WORKS(cppunit,[
-  LIBCPPUNIT_LIBS="$LIBCPPUNIT_PATH -lcppunit"
   PKG_CHECK_MODULES([LIBCPPUNIT],[cppunit],[
```

Comment:
@rousskov, It depends on which tests are being performed. The ones we can safely manipulate are: `CFLAGS, CPPFLAGS, CXXFLAGS, LDFLAGS, LIBS`

FTR; the full set of variables is documented at https://www.gnu.org/software/automake/manual/html_node/Program-and-Library-Variables.html. The subset which is "safe" is determined by our `SQUID_STATE_SAVE`/`SQUID_STATE_ROLLBACK` macros.

Reply:
> Alex: What is the best way [to adjust variables before calling PKG_CHECK_MODULES()]?

> Amos: It depends on which tests are being performed.

To avoid this complex and hidden dependency, can we _always_ adjust _all_ variables (that are safe to adjust) before calling PKG_CHECK_MODULES()?

If the answer is "yes", then we could, for example, add a simple new macro and use it in every PKG_CHECK_MODULES() caller (or perhaps even wrapping PKG_CHECK_MODULES to make these adjustments automatically?).

Reply:
The fact that catches my attention on this macro is that it is inconsistently applied.
When it is used, I would expect a repetitive pattern such as:
```
SQUID_AUTO_LIB(libfoo, [Example lib], LIBFOO)
SQUID_CHECK_LIB_WORKS(foo, [
  SQUID_STATE_SAVE(lfoo)
  AS_IF([test "x$LIBFOO_LIBS$LIBFOO_CFLAGS" = "x" ],[
    PKG_CHECK_MODULES([LIBFOO], [foo])
  ])
  dnl Do functionality tests
  CPPFLAGS="$CPPFLAGS $LIBFOO_CFLAGS"
  LIBS="$LIBS $LIBFOO_LIBS"
  AC_CHECK_HEADERS([foo.h foo/foo.h])
  AC_CHECK_LIB(foo, [foo_init])
  dnl logic drawing conclusions, possibly aborting with AC_MSG_ERROR or disabling features
  SQUID_STATE_ROLLBACK(lfoo)
])
```

But then:
```
  LIBS="$LIBS $LIBSYSTEMD_PATH"
  PKG_CHECK_MODULES(LIBSYSTEMD,[libsystemd],,[
    # systemd < 209
    PKG_CHECK_MODULES(LIBSYSTEMD,[libsystemd-daemon],,[:])
  ])
  AC_CHECK_HEADERS(systemd/sd-daemon.h)
```

AFAICT, PKG_CHECK_MODULES doesn't change or CPPFLAGS, LIBS, only MODULE_{CFLAGS,LIBS}, so if systemd is installed on a nonstandard location (unlikely, but still..) the check will fail unnecessarily.
Also, LIBSYSTEMD_PATH is only mentioned here from the whole source tree, it's somewhat confusing as well

In the test for `netfilter-conntrack`, state is not saved, and the feature is only disabled if the headers can't be found, not if the library can't be found

And so on and so forth, it's all very inconsistent.
Could we maybe converge on a repeated pattern such as the one above (after perfecting it, of course) and copypaste it?

Reply:
The fact that catches my attention on this macro is that it is inconsistently applied.
When it is used, I would expect a repetitive pattern such as:
```
SQUID_AUTO_LIB(libfoo, [Example lib], LIBFOO)
SQUID_CHECK_LIB_WORKS(foo, [
  SQUID_STATE_SAVE(lfoo)
  AS_IF([test "x$LIBFOO_LIBS$LIBFOO_CFLAGS" = "x" ],[
    PKG_CHECK_MODULES([LIBFOO], [foo])
  ])
  dnl Do functionality tests
  CPPFLAGS="$CPPFLAGS $LIBFOO_CFLAGS"
  LIBS="$LIBS $LIBFOO_LIBS"
  AC_CHECK_HEADERS([foo.h foo/foo.h])
  AC_CHECK_LIB(foo, [foo_init])
  dnl logic drawing conclusions, possibly aborting with AC_MSG_ERROR or disabling features
  SQUID_STATE_ROLLBACK(lfoo)
])
```

But then:
```
  LIBS="$LIBS $LIBSYSTEMD_PATH"
  PKG_CHECK_MODULES(LIBSYSTEMD,[libsystemd],,[
    # systemd < 209
    PKG_CHECK_MODULES(LIBSYSTEMD,[libsystemd-daemon],,[:])
  ])
  AC_CHECK_HEADERS(systemd/sd-daemon.h)
```

AFAICT, PKG_CHECK_MODULES doesn't change or CPPFLAGS, LIBS, only MODULE_{CFLAGS,LIBS}, so if systemd is installed on a nonstandard location (unlikely, but still..) the check will fail unnecessarily.
Also, LIBSYSTEMD_PATH is only mentioned here from the whole source tree, it's somewhat confusing as well

In the test for `netfilter-conntrack`, state is not saved, and the feature is only disabled if the headers can't be found, not if the library can't be found

And so on and so forth, it's all very inconsistent.
Could we maybe converge on a repeated pattern such as the one above (after perfecting it, of course) and copypaste it?

(I'm really confused where does the _PATH part comes from . SQUID_CHECK_LIB_WORKS mentions it as well, but it's really not clear where it's set)

Reply:
Okay, the proposal definitely needs some revisiting, I have reread SQUID_LIB_WORKS and it takes care of checking that the administrator has not set LIB_LIBS so that AS_IF is redundant. Still the idea of a template here seems worthwhile to me

Reply:
> Francesco: Could we maybe converge on a repeated pattern such as the one above (after perfecting it, of course)

That is the outcome I was hoping for when starting this thread. Once the correct pattern is identified, it can be codified (to avoid duplication that usually leads to bugs and other problems).

> Amos: Please leave that change out of scope for this PR.

I do not insist on any changes in this thread. Ad hoc fixes are fine, but it would be much better if those fixes were based on (and were moving us towards) that correct pattern. It looks like we still do not know what that pattern is, unfortunately. Glad you are working on developing that knowledge.